pkg/clusterresourceoverride/limitquerier.go

```diff
@@ -13,48 +13,66 @@ type namespaceLimitQuerier struct {
 	limitRangesLister corev1listers.LimitRangeLister
 }
 
-func (l *namespaceLimitQuerier) QueryMinimum(namespace string) (minimum *Floor, err error) {
-	limits, listErr := l.limitRangesLister.LimitRanges(namespace).List(labels.Everything())
+func (l *namespaceLimitQuerier) QueryFloorAndCeiling(namespace string) (floor *CPUMemory, ceiling *CPUMemory, err error) {
+	limitRanges, listErr := l.limitRangesLister.LimitRanges(namespace).List(labels.Everything())
 	if listErr != nil {
 		err = fmt.Errorf("failed to query limitrange - %v", listErr)
 		return
 	}
 
-	nsCPUFloor := minResourceLimits(limits, corev1.ResourceCPU)
-	nsMemFloor := minResourceLimits(limits, corev1.ResourceMemory)
+	nsCPUMinimum, nsCPUMaximum := GetMinMax(limitRanges, corev1.ResourceCPU)
+	nsMemMinimum, nsMemMaximum := GetMinMax(limitRanges, corev1.ResourceMemory)
 
-	minimum = &Floor{
-		CPU:    nsCPUFloor,
-		Memory: nsMemFloor,
+	floor = &CPUMemory{
+		CPU:    nsCPUMinimum,
+		Memory: nsMemMinimum,
+	}
+	ceiling = &CPUMemory{
+		CPU:    nsCPUMaximum,
+		Memory: nsMemMaximum,
 	}
 	return
 }
 
-// minResourceLimits finds the Min limit for resourceName. Nil is
-// returned if limitRanges is empty or limits contains no resourceName
-// limits.
-func minResourceLimits(limitRanges []*corev1.LimitRange, resourceName corev1.ResourceName) *resource.Quantity {
-	limits := []*resource.Quantity{}
+// GetMinMax finds the Minimum and Maximum limit for respectively for the specified resource.
+// Nil is returned if limitRanges is empty or limits contains no resourceName limits.
+func GetMinMax(limitRanges []*corev1.LimitRange, resourceName corev1.ResourceName) (minimum *resource.Quantity, maximum *resource.Quantity) {
+	minList, maxList := findMinMaxLimits(limitRanges, resourceName)
+
+	minimum = minQuantity(minList)
+	maximum = maxQuantity(maxList)
+
+	return
+}
+
+func findMinMaxLimits(limitRanges []*corev1.LimitRange, resourceName corev1.ResourceName) (minimum []*resource.Quantity, maximum []*resource.Quantity) {
+	minimum = []*resource.Quantity{}
+	maximum = []*resource.Quantity{}
 
 	for _, limitRange := range limitRanges {
-		for _, limit := range limitRange.Spec.Limits {
-			if limit.Type == corev1.LimitTypeContainer {
-				if limit, found := limit.Min[resourceName]; found {
-					clone := limit.DeepCopy()
-					limits = append(limits, &clone)
+		for _, limits := range limitRange.Spec.Limits {
+			if limits.Type == corev1.LimitTypeContainer {
+				if min, found := limits.Min[resourceName]; found {
+					clone := min.DeepCopy()
+					minimum = append(minimum, &clone)
+				}
+
+				if max, found := limits.Max[resourceName]; found {
+					clone := max.DeepCopy()
+					maximum = append(maximum, &clone)
 				}
 			}
 		}
 	}
 
-	if len(limits) == 0 {
-		return nil
-	}
-
-	return minQuantity(limits)
+	return
 }
 
 func minQuantity(quantities []*resource.Quantity) *resource.Quantity {
+	if len(quantities) == 0 {
+		return nil
+	}
+
```

Comment:
the deepcopies below you can optimize into one at the very end

Reply:
This is the original func ported from 3.x. Fixed it, although dereferrencing is almost equivalent to `DeepCopy` here.

Reply:
It was bad before :)